src/v/storage/readers_cache.cc

```diff
@@ -0,0 +1,252 @@
+    return evict_if(
+      [o](entry& e) { return e.reader->lease_range_end_offset() >= o; });
+}
+
+ss::future<> readers_cache::evict_range(model::offset base, model::offset end) {
+    vlog(
+      stlog.debug,
+      "{} - evicting reader from cache, range [{},{}]",
+      _ntp,
+      base,
+      end);
+    return evict_if([base, end](entry& e) {
+        return !(
+          e.reader->lease_range_base_offset() > end
+          || e.reader->lease_range_end_offset() < base);
+    });
+}
+
+model::record_batch_reader
+readers_cache::entry::make_cached_reader(readers_cache* cache) {
+    class cached_reader_impl final : public model::record_batch_reader::impl {
+    public:
+        explicit cached_reader_impl(entry* e, readers_cache* c)
+          : _underlying(e->reader.get())
+          , _guard(e, c) {}
+        cached_reader_impl(cached_reader_impl&&) noexcept = default;
+        cached_reader_impl& operator=(cached_reader_impl&&) noexcept = default;
+
+        cached_reader_impl(const cached_reader_impl&) noexcept = delete;
+        cached_reader_impl&
+        operator=(const cached_reader_impl&) noexcept = delete;
+
+        bool is_end_of_stream() const final {
+            return _underlying->is_end_of_stream();
+        };
+
+        ss::future<model::record_batch_reader::storage_t>
+        do_load_slice(model::timeout_clock::time_point tout) final {
+            return _underlying->do_load_slice(tout);
+        }
+
+        ss::future<> finally() noexcept final { return ss::now(); }
+
+        void print(std::ostream& o) final { return _underlying->print(o); };
+        ~cached_reader_impl() final = default;
+
+    private:
+        log_reader* _underlying;
+        entry_guard _guard;
+    };
+
+    return model::make_record_batch_reader<cached_reader_impl>(this, cache);
+}
+
+readers_cache::~readers_cache() {
+    vassert(
+      _readers.empty() && _in_use.empty(),
+      "readers cache have to be closed before destorying");
+}
+
+ss::future<>
+readers_cache::dispose_entries(intrusive_list<entry, &entry::_hook> entries) {
+    for (auto& e : entries) {
+        co_await e.reader->finally();
+    }
+
+    entries.clear_and_dispose([this](entry* e) {
+        vlog(
+          stlog.trace,
+          "{} - removing reader: [{},{}] lower_bound: {}",
+          _ntp,
+          e->reader->lease_range_base_offset(),
+          e->reader->lease_range_end_offset(),
+          e->reader->next_read_lower_bound());
+        delete e; // NOLINT
+    });
+}
+
+void readers_cache::dispose_in_background(
+  intrusive_list<entry, &entry::_hook> entries) {
+    (void)ss::with_gate(_gate, [this, entries = std::move(entries)]() mutable {
+        return dispose_entries(std::move(entries));
+    }).handle_exception_type([](const ss::gate_closed_exception&) {
+        // ignore, we are stopping cache
+    });
+    ;
+}
+
+void readers_cache::dispose_in_background(entry* e) {
+    (void)ss::with_gate(_gate, [this, e] {
```

Comment:
since an entry is in a reader wrapper owned by a independent fiber, can the reader cache shutdown asynchronously to the reader fiber destroying its reader and invoking this via the entry destructor, resulting in with_gate throwing?

Reply:
yes, I've added correct handling for `gate_closed_exception`